ModelBinders/GovUkMandatoryDateBinder.cs

```diff
@@ -0,0 +1,81 @@
+﻿using Microsoft.AspNetCore.Mvc.ModelBinding;
+using System.Linq;
+using System.Threading.Tasks;
+using GovUkDesignSystem.Attributes.DataBinding;
+using System;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Mvc.ModelBinding.Validation;
+
+namespace GovUkDesignSystem.ModelBinders
+{
+    /// <summary>
+    /// This model binder can be used to replace the default MVC model binder for a required date property. It will add
+    /// validation messages to the model state inline with the GovUk Design System guidelines.
+    /// This binder must be used alongside a GovUkDataBindingDateErrorTextAttribute attribute.
+    /// </summary>
+    public class GovUkMandatoryDateBinder : IModelBinder
+    {
+        public Task BindModelAsync(ModelBindingContext bindingContext)
+        {
+            var errorText = bindingContext.ModelMetadata.ValidatorMetadata.OfType<GovUkDataBindingDateErrorTextAttribute>().SingleOrDefault();
+            if (errorText == null)
+            {
+                throw new Exception("When using the GovUkMandatoryDateBinder you must also provide a GovUkDataBindingDateErrorTextAttribute attribute and ensure that you register GovUkDataBindingErrorTextProvider in your application's Startup.ConfigureServices method.");
+            }
+            var modelName = bindingContext.ModelName;
+            var modelNames = new[] { modelName + "-day", modelName + "-month", modelName + "-year" };
+            var modelValueDictionary = new List<KeyValuePair<string, ValueProviderResult>>();
+
+            modelNames.ToList().ForEach(m => { modelValueDictionary.Add(new KeyValuePair<string, ValueProviderResult>(m, bindingContext.ValueProvider.GetValue(m))); });
+
+            // Ensure that a not empty value was sent to us in the request
+            if (modelValueDictionary.Any(r => r.Value == ValueProviderResult.None || string.IsNullOrEmpty(r.Value.FirstValue))) 
+            {
+                return Task.CompletedTask;
+            }
+
+            var valuesDictionary = modelValueDictionary.Select( v => new KeyValuePair<string, string>(v.Key,v.Value.FirstValue)).ToList();
+
+            var errors = new List<KeyValuePair<string, DateErrors>>();
+            valuesDictionary.ForEach(p =>
+            {
+                if (string.IsNullOrEmpty(p.Value))
+                {
+                    var unit = p.Key.Substring(p.Key.LastIndexOf("-") + 1, p.Key.Length - p.Key.LastIndexOf("-") - 1);
+                    errors.Add(new KeyValuePair<string, DateErrors>(unit, DateErrors.ValueMissing));
+                }
+            });
+
+            if (errors.Count == 0)
+            {
+                var day = Int32.Parse(valuesDictionary[0].Value);
```

Comment:
This will throw an exception if someone enters a non-numeric value. You may want Int32.TryParse()